fs: public folder redesign, except for avatars

jakob223 opened the pull request:
Issue: KBFS-4172

songgao commented on shared/constants/fs.js:
```diff
@@ -601,6 +595,17 @@ export const getTlfFromTlfs = (tlfs: Types.Tlfs, tlfType: Types.TlfType, name: s
+    default:
+      return false
+  }
+}
```
IIRC "your own public TLF" means you're the only writer there, so just `.contains` wouldn't suffice.

But perhaps we should show it in all public TLFs? @cecileboucheron wdyt?

jakob223 replied:
I think we should show it in any public TLFs you have edit access to - hence `contains`. But interested to hear what Cécile thinks.

songgao replied:
Ah I see. In that case you could also use `writable` in the pathItem, which covers edge cases like when you access it through a social assertion handle as yourself.

That aside, it seems to me this would be a place to educate user about these stuff being open to the world, _and_ to let them know about keybase.pub, which is why I thought maybe we should show it for all. Design has final call though!

cecileboucheron replied:
Ah yeah good call, let's show it in all the public TLFs where you're a writer. So shared public folders too.

songgao commented on shared/fs/banner/public-reminder.js:
```diff
@@ -0,0 +1,47 @@
+  state = {closed: false}
+
+  _onClose = () => this.setState({closed: true})
+
```
I think this works only because we never have scenario where we go directly from one public TLF to another -- in which case this component will get be kept mounted and maintain the closed state, thus fail to bring back the banner. Seems we could have a `componentDidUpdate` here to check if path change results in TLF change, just to be safe.

We could also add something like "lastDismissedPublicBannerTlf" in the redux store, depending on when exactly the banner is supposed to come back. For example, should it come back if user is in the Files tab, goes to chat, and comes back to the Files tab and land in the same public folder? Should it come back if user goes back out to the TLF list from a public folder, then back into the same public folder?

jakob223 replied:
Hmm, this is a good point. I'll put it in the redux store.

> For example, should it come back if user is in the Files tab, goes to chat, and comes back to the Files tab and land in the same public folder?   
I guess no here.

> Should it come back if user goes back out to the TLF list from a public folder, then back into the same public folder?
Not sure here, I'd be okay with yes. WDYT?

songgao replied:
Not sure! @cecileboucheron wdyt?

jakob223 replied:
hmm, so upon trying to implement this I am finding that the redux solution adds a lot of complication and doesn't actually seem to yield a consistent solution that I can think of. I think I'm going to go with the simpler state version with `componentDidUpdate` for now.

songgao replied:
I think it still depends on the behavior we want. If we want the "dismissed" state persist through screens, using the state wouldn't work as we'd lose states after unmount happens.

songgao commented on shared/fs/banner/public-reminder.js:
```diff
@@ -0,0 +1,47 @@
+      <Kb.Banner
+        color="yellow"
+        text="Everything you upload in here can be viewed by everyone at "
+        actions={[{onClick: () => openUrl(url), title: url + '.'}]}
```
If you do [the thing that determines if the banner should show up] inside this component, you'd never need to worry about showing his while url is `''`. I don't feel very strongly about it though so up to you!

jakob223 commented:
Untagging review while I fix a mobile thing..

jakob223 commented:
Ok, @songgao, I've rebased this on master, converted to TS, and fixed the behavior of the close button to what you described above. Can you take another look when you have a chance?

songgao commented on shared/fs/banner/public-reminder.tsx:
```diff
@@ -0,0 +1,83 @@
+          actions={[
+            {
+              onClick: () => openUrl(props.url),
+              title: props.url + '.',
```
Why the `.` in the end? That'd become part of the underlined clickable.

jakob223 replied:
Yeah, but there's no way to include a `.` after the actions with `Banner`, and there's a `.` in the design.

songgao commented on shared/fs/banner/public-reminder.tsx:
```diff
@@ -0,0 +1,83 @@
+
+type DispatchProps = {_onClose: (string) => () => void}
+const mapDispatchToProps = dispatch => ({
+  _onClose: tlf => () => dispatch(FsGen.createSetLastPublicBannerClosedTlf({tlf})),
```
Could also compute `tlf` here from `ownProps`. I suppose you did it this way so `parsePath` is only called once? Perhaps we could memoize-one `parsePath`.

jakob223 replied:
also, reusing `_onClose` to clear the state. How about renaming `_onClose` to `setLastPublicBannerClosedTlf` or something?

songgao commented on shared/fs/banner/public-reminder.tsx:
```diff
@@ -0,0 +1,83 @@
+      )
+    )
+  } else {
+    props.onLoad()
```
Yeah, this needs to be in lifecycles or `useEffect` rather than `render()`.

An alternative is to have the connector pass in simpler actions (i.e. `onSetLastClosed` and `clearLastClosed`), and have this component whether we should set or clear or do nothing.

songgao commented on shared/fs/banner/public-reminder.tsx:
```diff
@@ -0,0 +1,83 @@
+      return {
+        hidden: s._lastClosedTlf !== parsedPath.tlfName,
+        onClose: d._onClose(parsedPath.tlfName),
+        onLoad: s._lastClosedTlf === '' ? () => {} : d._onClose(''),
```
This would cause the banner to come back when you navigate inside a public TLF too right? Seems like we'd need to compare against this with current tlfName too.

jakob223 replied:
This only gets called when `show` is false. Maybe rename to `onLoadNonBannerFolder` or something?

songgao commented on shared/fs/nav-header/mobile-header.tsx:
```diff
@@ -16,6 +16,22 @@ type State = {
+    default:
+  }
+  return false
+}
```
There's also `Constants.getPathVisibility`

jakob223 replied:
😮

songgao commented on shared/fs/nav-header/title.tsx:
```diff
@@ -84,10 +84,29 @@ const Breadcrumb = Kb.OverlayParentHOC(
+    default:
+  }
+  return null
+}
```
Same here. If you prefer to use `showPublicTag` above, it should probably go into `constants/fs` so it can be used here too.

jakob223 commented on shared/common-adapters/avatar-line.tsx:
```diff
@@ -0,0 +1,96 @@
+            style={styles.avatar}
+          />
+        ))
+        .reverse()}
```
This is so that the last avatar is underneath the first one.
